thesis/inc/technical/detailed_architecture/component_collection.tex

```diff
@@ -0,0 +1,135 @@
+\subsection{Component Collection}
+\label{subsec:detailed_component_collection}
+The component collection is where all the different building blocks
+comes together.
+As the name suggests the component collection is responsible for containing
+a collection of a specific component type.
+The collection was design for fast contiguous processing of component groups internally within the ECS,
+while giving sufficient to external lookups on individual components.
+
+\subsubsection{Explanation}
+The component collection functions like a dynamically sized pool for components of a specific type.
+The following paragraphs explains the different decisions faced when creating the component collection.
+
+\paragraph{Contiguous Byte Buffer}
+Components are stored in a contiguous buffer of raw memory,
+which is manually iterated over with byte pointers.
+The rationale behind this is that the type of a component is opaque
+to the collection, meaning it is unknown at compile time
+how many bytes that is needed to increment a pointer to the next object.
+Operations on the components are done by reinterpreting the components from bytes to components,
+and send them as parameters to the operation types belonging to the component.
+The reasoning for using a contiguous memory buffer is to make better usage of hardware caching mechanisms.
+
+\paragraph{Reallocation \& Growth Strategy}
+As mentioned the buffer is contiguous memory, which needs to be reallocated when new components are added.
+this can be a costly operation, but is mitigated by using a simple growth strategy.
+When a collection needs to reallocate it will expand the buffer size based it's growth factor.
+The new size of the collection is simply its old size multiplied with its growth factor.
+This leads to an amortized constant complexity when adding a new component to the collection,
+however some extra overhead is required to keep the index map in a valid state.
+
+\paragraph{Lifecycle Partitioning}
+\todo{Add Figure: component\_collection.png}
+The components in the collection is further grouped together and partitioned based on their stage
+in the entity lifecycle\secref{subsec:high_level_lifecycle}.
+Pointers are used to indicate where the different partitions in the collection begin.
+Transitioning components to other stages in the lifecycle therefore means to find the component,
+swapping it with the the first component in a partition and increment or decrement the partition pointer.
+
+\paragraph{Index Mapping}
+Collections are often queried for their components, meaning that some form of lookup is needed.
+Several approaches could have been taken to solve this problem,
+the path taken in the collection was to add an indexing structure on top of the collection.
```

Comment:
"Several approaches could have been taken to solve this problem, the path taken in the collection" seems a bit over the top, and not really necessary. Would replace it with "The problem was resolved by adding an indexing structure..."